File: swift/swiftfind/Sources/swiftfind/FindOptions.swift
//
//  FindOptions.swift
//  swiftfind
//
//  Created by Cary Clark on 5/17/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Foundation

public protocol Option {
    var shortArg: String? { get }
    var longArg: String { get }
    var desc: String { get }
    var argType: ArgTokenType { get }
}

struct FindOption: Option {
    let shortArg: String?
    let longArg: String
    let desc: String
    let argType: ArgTokenType

    var sortArg: String {
        if shortArg != nil, !shortArg!.isEmpty {
            return shortArg!.lowercased() + "a" + longArg.lowercased()
        }
        return longArg.lowercased()
    }
}

public class FindOptions {
    private var config: FindConfig
    private var findOptions = [FindOption]()
    private var argTokenizer: ArgTokenizer?

    public init() {
        config = FindConfig()
        setFindOptionsFromJson()
        argTokenizer = ArgTokenizer(findOptions)
    }

    private func setFindOptionsFromJson() {
        do {
            let findOptionsUrl = URL(fileURLWithPath: config.findOptionsPath)
            let data = try Data(contentsOf: findOptionsUrl, options: .mappedIfSafe)
            if let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
                if let options = json["findoptions"] as? [[String: Any]] {
                    for so in options {
                        let longArg = so["long"] as! String
                        let shortArg = so.index(forKey: "short") != nil ? (so["short"] as! String) : nil
                        let desc = so["desc"] as! String
                        var argType = ArgTokenType.unknown
                        if self.boolActionDict.index(forKey: longArg) != nil {
                            argType = ArgTokenType.bool
                        } else if self.stringActionDict.index(forKey: longArg) != nil {
                            argType = ArgTokenType.str
                        } else if self.intActionDict.index(forKey: longArg) != nil {
                            argType = ArgTokenType.int
                        } else if self.longActionDict.index(forKey: longArg) != nil {
                            argType = ArgTokenType.long
                        }
                        findOptions.append(FindOption(shortArg: shortArg, longArg: longArg, desc: desc, argType: argType))
                    }
                }
            }
        } catch let error as NSError {
            print("Failed to load: \(error.localizedDescription)")
        }
    }

    private let boolActionDict: [String: (Bool, FindSettings) -> Void] = [
        "archivesonly": { (bool: Bool, settings: FindSettings) in
            settings.archivesOnly = bool
        },
        "colorize": { (bool: Bool, settings: FindSettings) in
            settings.colorize = bool
        },
        "debug": { (bool: Bool, settings: FindSettings) in
            settings.debug = bool
        },
        "excludearchives": { (bool: Bool, settings: FindSettings) in
            settings.includeArchives = !bool
        },
        "excludehidden": { (bool: Bool, settings: FindSettings) in
            settings.includeHidden = !bool
        },
        "followsymlinks": { (bool: Bool, settings: FindSettings) in
            settings.followSymlinks = bool
        },
        "help": { (bool: Bool, settings: FindSettings) in
            settings.printUsage = bool
        },
        "includearchives": { (bool: Bool, settings: FindSettings) in
            settings.includeArchives = bool
        },
        "includehidden": { (bool: Bool, settings: FindSettings) in
            settings.includeHidden = bool
        },
        "nocolorize": { (bool: Bool, settings: FindSettings) in
            settings.colorize = !bool
        },
        "nofollowsymlinks": { (bool: Bool, settings: FindSettings) in
            settings.followSymlinks = !bool
        },
        "noprintdirs": { (bool: Bool, settings: FindSettings) in
            settings.printDirs = !bool
        },
        "noprintfiles": { (bool: Bool, settings: FindSettings) in
            settings.printFiles = !bool
        },
        "norecursive": { (bool: Bool, settings: FindSettings) in
            settings.recursive = !bool
        },
        "printdirs": { (bool: Bool, settings: FindSettings) in
            settings.printDirs = bool
        },
        "printfiles": { (bool: Bool, settings: FindSettings) in
            settings.printFiles = bool
        },
        "recursive": { (bool: Bool, settings: FindSettings) in
            settings.recursive = bool
        },
        "sort-ascending": { (bool: Bool, settings: FindSettings) in
            settings.sortDescending = !bool
        },
        "sort-caseinsensitive": { (bool: Bool, settings: FindSettings) in
            settings.sortCaseInsensitive = bool
        },
        "sort-casesensitive": { (bool: Bool, settings: FindSettings) in
            settings.sortCaseInsensitive = !bool
        },
        "sort-descending": { (bool: Bool, settings: FindSettings) in
            settings.sortDescending = bool
        },
        "verbose": { (bool: Bool, settings: FindSettings) in
            settings.verbose = bool
        },
        "version": { (bool: Bool, settings: FindSettings) in
            settings.printVersion = bool
        }
    ]

    private let stringActionDict: [String: (String, FindSettings) -> Void] = [
        "in-archiveext": { (str: String, settings: FindSettings) in
            settings.addInArchiveExtension(str)
        },
        "in-archivefilepattern": { (str: String, settings: FindSettings) in
            settings.addInArchiveFilePattern(str)
        },
        "in-dirpattern": { (str: String, settings: FindSettings) in
            settings.addInDirPattern(str)
        },
        "in-ext": { (str: String, settings: FindSettings) in
            settings.addInExtension(str)
        },
        "in-filepattern": { (str: String, settings: FindSettings) in
            settings.addInFilePattern(str)
        },
        "in-filetype": { (str: String, settings: FindSettings) in
            settings.addInFileType(str)
        },
        "out-archiveext": { (str: String, settings: FindSettings) in
            settings.addOutArchiveExtension(str)
        },
        "maxlastmod": { (str: String, settings: FindSettings) in
            settings.setMaxLastModFromString(str)
        },
        "minlastmod": { (str: String, settings: FindSettings) in
            settings.setMinLastModFromString(str)
        },
        "out-archivefilepattern": { (str: String, settings: FindSettings) in
            settings.addOutArchiveFilePattern(str)
        },
        "out-dirpattern": { (str: String, settings: FindSettings) in
            settings.addOutDirPattern(str)
        },
        "out-ext": { (str: String, settings: FindSettings) in
            settings.addOutExtension(str)
        },
        "out-filepattern": { (str: String, settings: FindSettings) in
            settings.addOutFilePattern(str)
        },
        "out-filetype": { (str: String, settings: FindSettings) in
            settings.addOutFileType(str)
        },
        "path": { (str: String, settings: FindSettings) in
            settings.addPath(str)
        },
        "sort-by": { (str: String, settings: FindSettings) in
            settings.setSortBy(str)
        }
    ]

    private let intActionDict: [String: (Int32, FindSettings) -> Void] = [
        "maxdepth": { (i: Int32, settings: FindSettings) in
            settings.maxDepth = i
        },
        "mindepth": { (i: Int32, settings: FindSettings) in
            settings.minDepth = i
        },
    ]

    private let longActionDict: [String: (UInt64, FindSettings) -> Void] = [
        "maxsize": { (l: UInt64, settings: FindSettings) in
            settings.maxSize = l
        },
        "minsize": { (l: UInt64, settings: FindSettings) in
            settings.minSize = l
        },
    ]

    public func updateSettingsFromArgTokens(_ settings: FindSettings, argTokens: [ArgToken]) throws {
        for argToken in argTokens {
            if argToken.type == ArgTokenType.bool {
                if let bool = argToken.value as? Bool {
                    boolActionDict[argToken.name]!(bool, settings)
                } else {
                    throw FindError(msg: "Invalid value for option: \(argToken.name)")
                }
            } else if argToken.type == ArgTokenType.str {
                if let string = argToken.value as? String {
                    if argToken.name == "settings-file" {
                        try updateSettingsFromFile(settings, filePath: string)
                    } else {
                        stringActionDict[argToken.name]!(string, settings)
                    }
                } else if let stringArray = argToken.value as? [String] {
                    for s in stringArray {
                        stringActionDict[argToken.name]!(s, settings)
                    }
                } else {
                    throw FindError(msg: "Invalid value for option: \(argToken.name)")
                }
            } else if argToken.type == ArgTokenType.int {
                if let intVal = argToken.value as? Int32 {
                    intActionDict[argToken.name]!(intVal, settings)
                } else {
                    throw FindError(msg: "Invalid value for option: \(argToken.name)")
                }
            }  else if argToken.type == ArgTokenType.long {
                if let longVal = argToken.value as? UInt64 {
                    longActionDict[argToken.name]!(longVal, settings)
                } else {
                    throw FindError(msg: "Invalid value for option: \(argToken.name)")
                }
            } else {
                throw FindError(msg: "Invalid option: \(argToken.name)")
            }
        }
    }

    public func updateSettingsFromJson(_ settings: FindSettings, jsonString: String) throws {
        let argTokens = try argTokenizer!.tokenizeJson(jsonString)
        try updateSettingsFromArgTokens(settings, argTokens: argTokens)
    }

    public func settingsFromJson(_ jsonString: String) throws -> FindSettings {
        let settings = FindSettings()
        try updateSettingsFromJson(settings, jsonString: jsonString)
        return settings
    }

    public func updateSettingsFromFile(_ settings: FindSettings, filePath: String) throws {
        let argTokens = try argTokenizer!.tokenizeFile(filePath)
        try updateSettingsFromArgTokens(settings, argTokens: argTokens)
    }

    public func settingsFromFile(_ filePath: String) throws -> FindSettings {
        let settings = FindSettings()
        try updateSettingsFromFile(settings, filePath: filePath)
        return settings
    }

    public func updateSettingsFromArgs(_ settings: FindSettings, args: [String]) throws {
        let argTokens = try argTokenizer!.tokenizeArgs(args)
        try updateSettingsFromArgTokens(settings, argTokens: argTokens)
    }

    public func settingsFromArgs(_ args: [String]) throws -> FindSettings {
        let settings = FindSettings()
        // default printFiles to true since running in cli
        settings.printFiles = true
        try updateSettingsFromArgs(settings, args: args)
        return settings
    }

    func getUsageString() -> String {
        var str = "\nUsage:\n swiftfind [options] <path> [<path> ...]\n\n"
        str += "Options:\n"
        let options = findOptions.sorted(by: { $0.sortArg < $1.sortArg })

        let optStrings = options.map {
            switch ($0.shortArg, $0.longArg) {
            // Order errors by code
            case let (nil, longArg):
                "--\(longArg)"
            case let ("", longArg):
                "--\(longArg)"
            case let (shortArg, longArg):
                "-\(shortArg!),--\(longArg)"
            }
        }

        let optDescs = options.map(\.desc)
        let longest = optStrings.map { $0.lengthOfBytes(using: String.Encoding.utf8) }.max()!
        for i in 0 ..< optStrings.count {
            var optLine = " \(optStrings[i])"
            while optLine.lengthOfBytes(using: String.Encoding.utf8) <= longest {
                optLine += " "
            }
            optLine += "  \(optDescs[i])\n"
            str += optLine
        }
        return str
    }

    public func usage(_ code: Int32 = 0) {
        logMsg(getUsageString())
        exit(code)
    }
}

File: csharp/CsFind/CsFindLib/FindOptions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using FindOptionsDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<System.Collections.Generic.Dictionary<string, string>>>;

namespace CsFindLib;

public class FindOptions
{
	private readonly string _findOptionsResource;

	private static readonly Dictionary<string, Action<bool, FindSettings>> BoolActionDictionary =
		new()
		{
			{ "archivesonly", (b, settings) => settings.ArchivesOnly = b },
			{ "colorize", (b, settings) => settings.Colorize = b },
			{ "debug", (b, settings) => settings.Debug = b },
			{ "excludearchives", (b, settings) => settings.IncludeArchives = !b },
			{ "excludehidden", (b, settings) => settings.IncludeHidden = !b },
			{ "followsymlinks", (b, settings) => settings.FollowSymlinks = b },
			{ "help", (b, settings) => settings.PrintUsage = b },
			{ "includearchives", (b, settings) => settings.IncludeArchives = b },
			{ "includehidden", (b, settings) => settings.IncludeHidden = b },
			{ "nocolorize", (b, settings) => settings.Colorize = !b },
			{ "nofollowsymlinks", (b, settings) => settings.FollowSymlinks = !b },
			{ "noprintdirs", (b, settings) => settings.PrintDirs = !b },
			{ "noprintfiles", (b, settings) => settings.PrintFiles = !b },
			{ "printdirs", (b, settings) => settings.PrintDirs = b },
			{ "printfiles", (b, settings) => settings.PrintFiles = b },
			{ "norecursive", (b, settings) => settings.Recursive = !b },
			{ "recursive", (b, settings) => settings.Recursive = b },
			{ "sort-ascending", (b, settings) => settings.SortDescending = !b },
			{ "sort-caseinsensitive", (b, settings) => settings.SortCaseInsensitive = b },
			{ "sort-casesensitive", (b, settings) => settings.SortCaseInsensitive = !b },
			{ "sort-descending", (b, settings) => settings.SortDescending = b },
			{ "verbose", (b, settings) => settings.Verbose = b },
			{ "version", (b, settings) => settings.PrintVersion = b },
		};

	private static readonly Dictionary<string, Action<string, FindSettings>> StringActionDictionary =
		new()
		{
			{ "in-archiveext", (s, settings) => settings.AddInArchiveExtension(s) },
			{ "in-archivefilepattern", (s, settings) => settings.AddInArchiveFilePattern(s) },
			{ "in-dirpattern", (s, settings) => settings.AddInDirPattern(s) },
			{ "in-ext", (s, settings) => settings.AddInExtension(s) },
			{ "in-filepattern", (s, settings) => settings.AddInFilePattern(s) },
			{ "in-filetype", (s, settings) => settings.AddInFileType(s) },
			{ "maxlastmod", (s, settings) => {
					settings.MaxLastMod = DateTime.Parse(s);
				}
			},
			{ "minlastmod", (s, settings) => {
					settings.MinLastMod = DateTime.Parse(s);
				}
			},
			{ "out-archiveext", (s, settings) => settings.AddOutArchiveExtension(s) },
			{ "out-archivefilepattern", (s, settings) => settings.AddOutArchiveFilePattern(s) },
			{ "out-dirpattern", (s, settings) => settings.AddOutDirPattern(s) },
			{ "out-ext", (s, settings) => settings.AddOutExtension(s) },
			{ "out-filepattern", (s, settings) => settings.AddOutFilePattern(s) },
			{ "out-filetype", (s, settings) => settings.AddOutFileType(s) },
			{ "path", (s, settings) => settings.AddPath(s) },
			{ "sort-by", (s, settings) => settings.SetSortBy(s) },
		};

	private static readonly Dictionary<string, Action<int, FindSettings>> IntActionDictionary =
		new()
		{
			{ "maxdepth", (i, settings) => settings.MaxDepth = i },
			{ "maxsize", (i, settings) => settings.MaxSize = i },
			{ "mindepth", (i, settings) => settings.MinDepth = i },
			{ "minsize", (i, settings) => settings.MinSize = i },
		};

	public List<IOption> Options { get; }
	private Dictionary<string, string> BoolDictionary { get; }
	private Dictionary<string, string> StringDictionary { get; }
	private Dictionary<string, string> IntDictionary { get; }
	private ArgTokenizer ArgTokenizer { get; }


	public FindOptions()
	{
		_findOptionsResource = EmbeddedResource.GetResourceFileContents("CsFindLib.Resources.findoptions.json");
		Options = [];
		BoolDictionary = new Dictionary<string, string>();
		StringDictionary = new Dictionary<string, string> { { "path", "path" } };
		IntDictionary = new Dictionary<string, string>();
		SetOptionsFromJson();
		ArgTokenizer = new ArgTokenizer(Options);
	}

	private void SetOptionsFromJson()
	{
		var findOptionsDict = JsonSerializer.Deserialize<FindOptionsDictionary>(_findOptionsResource);
		if (findOptionsDict == null
		    || !findOptionsDict.TryGetValue("findoptions", out List<Dictionary<string, string>>? optionDicts))
		{
			throw new FindException("Missing or invalid search options resource");
		}

		foreach (var optionDict in optionDicts)
		{
			var longArg = optionDict["long"];
			ArgTokenType argType;
			if (BoolActionDictionary.ContainsKey(longArg))
			{
				argType = ArgTokenType.Bool;
			}
			else if (StringActionDictionary.ContainsKey(longArg) || longArg.Equals("settings-file"))
			{
				argType = ArgTokenType.String;
			}
			else if (IntActionDictionary.ContainsKey(longArg))
			{
				argType = ArgTokenType.Int;
			}
			else
			{
				throw new FindException($"Invalid option: {longArg}");
			}
			string? shortArg = null;
			if (optionDict.TryGetValue("short", out var shortVal))
			{
				shortArg = shortVal;
			}
			var desc = optionDict["desc"];
			var option = new FindOption(shortArg, longArg, desc, argType);
			Options.Add(option);
		}
	}

	private void ApplyArgTokenToSettings(ArgToken argToken, FindSettings settings)
	{
		if (argToken.Type == ArgTokenType.Bool)
		{
			if (BoolActionDictionary.TryGetValue(argToken.Name, out var boolAction))
			{
				if (argToken.Value is bool b)
				{
					boolAction(b, settings);
				}
				else if (argToken.Value is JsonElement { ValueKind: JsonValueKind.False or JsonValueKind.True } jsonElem)
				{
					boolAction(jsonElem.GetBoolean(), settings);
				}
				else
				{
					throw new FindException($"Invalid value for option: {argToken.Name}");
				}
			}
			else
			{
				throw new FindException($"Invalid option: {argToken.Name}");
			}
		}
		else if (argToken.Type == ArgTokenType.String)
		{
			if (StringActionDictionary.TryGetValue(argToken.Name, out var stringAction))
			{
				if (argToken.Value is string s)
				{
					stringAction(s, settings);
				}
				else if (argToken.Value is IEnumerable<string> enumerable)
				{
					foreach (var enumVal in enumerable)
					{
						stringAction(enumVal, settings);
					}
				}
				else if (argToken.Value is JsonElement { ValueKind: JsonValueKind.String }  jsonElem)
				{
					var jsonString = jsonElem.GetString();
					if (jsonString is not null)
					{
						stringAction(jsonString, settings);
					}
					else
					{
						throw new FindException($"Invalid value for option: {argToken.Name}");
					}
				}
				else if (argToken.Value is JsonElement { ValueKind: JsonValueKind.Array }  jsonArr)
				{
					foreach (var arrVal in jsonArr.EnumerateArray())
					{
						if (arrVal is JsonElement { ValueKind: JsonValueKind.String }  jsonArrElem)
						{
							var jsonStr = jsonArrElem.GetString();
							if (jsonStr is not null)
							{
								stringAction(jsonStr, settings);
							}
							else
							{
								throw new FindException($"Invalid value for option: {argToken.Name}");
							}
						}
					}
				}
				else
				{
					throw new FindException($"Invalid value for option: {argToken.Name}");
				}
			}
			else if (argToken.Name.Equals("settings-file"))
			{
				if (argToken.Value is string settingsFilePath)
				{
					UpdateSettingsFromFile(settings, settingsFilePath);
				}
				else if (argToken.Value is IEnumerable<string> enumerable)
				{
					foreach (var enumVal in enumerable)
					{
						UpdateSettingsFromFile(settings, enumVal);
					}
				}
				else
				{
					throw new FindException($"Invalid value for option: {argToken.Name}");
				}
			}
			else
			{
				throw new FindException($"Invalid option: {argToken.Name}");
			}
		}
		else if (argToken.Type == ArgTokenType.Int)
		{
			if (IntActionDictionary.TryGetValue(argToken.Name, out var intAction))
			{
				if (argToken.Value is int i)
				{
					intAction(i, settings);
				}
				else if (argToken.Value is JsonElement { ValueKind: JsonValueKind.Number } jsonElem)
				{
					intAction(jsonElem.GetInt32(), settings);
				}
				else
				{
					throw new FindException($"Invalid value for option: {argToken.Name}");
				}
			}
		}
		else
		{
			throw new FindException($"Invalid option: {argToken.Name}");
		}
	}

	private void UpdateSettingsFromArgTokens(FindSettings settings, List<ArgToken> argTokens)
	{
		foreach (var argToken in argTokens)
		{
			ApplyArgTokenToSettings(argToken, settings);
		}
	}

	private void UpdateSettingsFromDictionary(FindSettings settings, Dictionary<string, object> dictionary)
	{
		var argTokens = ArgTokenizer.TokenizeDictionary(dictionary);
		UpdateSettingsFromArgTokens(settings, argTokens);
	}

	public void UpdateSettingsFromJson(FindSettings settings, string jsonString)
	{
		var argTokens = ArgTokenizer.TokenizeJson(jsonString);
		UpdateSettingsFromArgTokens(settings, argTokens);
	}

	public FindSettings SettingsFromJson(string jsonString)
	{
		var settings = new FindSettings();
		UpdateSettingsFromJson(settings, jsonString);
		return settings;
	}

	public void UpdateSettingsFromFile(FindSettings settings, string filePath)
	{
		var argTokens = ArgTokenizer.TokenizeFile(filePath);
		UpdateSettingsFromArgTokens(settings, argTokens);
	}

	public FindSettings SettingsFromFile(string filePath)
	{
		var settings = new FindSettings();
		UpdateSettingsFromFile(settings, filePath);
		return settings;
	}

	public void UpdateSettingsFromArgs(FindSettings settings, IEnumerable<string> args)
	{
		var argTokens = ArgTokenizer.TokenizeArgs(args);
		UpdateSettingsFromArgTokens(settings, argTokens);
	}

	public FindSettings SettingsFromArgs(IEnumerable<string> args)
	{
		// default to PrintFiles = true since this is called from CLI
		var settings = new FindSettings {PrintFiles = true};
		UpdateSettingsFromArgs(settings, args);
		return settings;
	}

	private string GetUsageString()
	{
		var sb = new StringBuilder();
		sb.AppendLine("\nUsage:");
		sb.AppendLine(" csfind [options] <path> [<path> ...]\n");
		sb.AppendLine("Options:");
		var optStrings = new List<string>();
		var optDescs = new List<string>();
		var longest = 0;
		foreach (var opt in Options.OrderBy(o => o.SortArg))
		{
			var optString = new StringBuilder();
			if (!string.IsNullOrWhiteSpace(opt.ShortArg))
			{
				optString.Append($"-{opt.ShortArg},");
			}
			optString.Append($"--{opt.LongArg}");
			if (optString.Length > longest)
			{
				longest = optString.Length;
			}
			optStrings.Add(optString.ToString());
			optDescs.Add(opt.Description);
		}
		var format = " {0,-" + longest + "}  {1}";
		for (var i = 0; i < optStrings.Count; i++)
		{
			sb.AppendLine(string.Format(format, optStrings[i], optDescs[i]));
		}
		return sb.ToString();
	}

	public void Usage(int exitCode = 0)
	{
		Console.WriteLine(GetUsageString());
		Environment.Exit(exitCode);
	}
}
